modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java

```diff
@@ -156,36 +153,13 @@ public String getType() {
 
     @Override
     public <T> T compile(String scriptName, String scriptSource, ScriptContext<T> context, Map<String, String> params) {
-        // classloader created here
-        final SecurityManager sm = System.getSecurityManager();
-        SpecialPermission.check();
-        Expression expr = AccessController.doPrivileged(new PrivilegedAction<Expression>() {
-            @Override
-            public Expression run() {
-                try {
-                    // snapshot our context here, we check on behalf of the expression
-                    AccessControlContext engineContext = AccessController.getContext();
-                    ClassLoader loader = getClass().getClassLoader();
-                    if (sm != null) {
-                        loader = new ClassLoader(loader) {
-                            @Override
-                            protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
-                                try {
-                                    engineContext.checkPermission(new ClassPermission(name));
-                                } catch (SecurityException e) {
-                                    throw new ClassNotFoundException(name, e);
-                                }
-                                return super.loadClass(name, resolve);
-                            }
-                        };
-                    }
-                    // NOTE: validation is delayed to allow runtime vars, and we don't have access to per index stuff here
-                    return JavascriptCompiler.compile(scriptSource, JavascriptCompiler.DEFAULT_FUNCTIONS, loader);
-                } catch (ParseException e) {
-                    throw convertToScriptException("compile error", scriptSource, scriptSource, e);
-                }
-            }
-        });
```

Comment:
I assume the security manager stuff is no longer needed after https://github.com/apache/lucene/pull/12873? I don't understand it but am sure something would break if we still needed it.

Reply:
I think the SpecialPermission check should still be present. https://github.com/apache/lucene/pull/12873 removes the need for the doPrivileged since we no longer create a classloader. But the SpecialPermission check is to ensure we are not trying to invoke this from a script.

Reply:
Yep, no longer needed. I discussed this with @jdconrad and @uschindler kindly took a look at this as well. The security manager code revolved around the class loader code needed to unload classes, which is now done differently using hidden classes that don't need manual unloading.

Reply:
Thanks for pointing this out. Correct, the class loader stuff is now replaced by using hidden classes, and the security manager code around class loader can also be removed. I did overlook the `SpecialPermission.check` line and removed it by mistake.

Reply:
I restored it here: https://github.com/elastic/elasticsearch/pull/114741/commits/b6f646c2b32f2af01ba320807db31574ade58c13 . 

Anything else I may have missed? :)

Reply:
Hi,
Actually the security manager extra hack with the classes wasn't really a must previously, because the compiler for expressions had a hard coded list of functions to call. Because Elasticsearch never added any own extensions and just passed DEFAULT_FUNCTIONS, there was never any risk.
With the new code it is more strict: at runtime it never loads any classes or calls any function directly anymore. All functions are Method handles which were previously looked up and linked. At runtime of expression it just involves the previously resolved method handles based on dynamic constants feature of JVM (LDC opcode to find Method handle based on name and then invokeExact). Because everything is resolved early, any script won't load any classes.
If Elasticsearch wants to add own functions, those must be looked up and linked to a MethodHandle in the plugin code and stored in the function map. So security manager or class loader checks would only be needed when building the method handles to call the functions later. At runtime of scripts all is just a dynamic constant LDC.

Reply:
Painless should also switch to hidden classes and dynamic constants to call methods at some point (P.S. most java lambdas use anonymous classes and use dynamic constants to call the target, too).

Reply:
Yeah I want to switch painless to hidden classes, just haven't got around to it:
https://github.com/elastic/elasticsearch/issues/102955